Spam Classification Model

Creating model to find user entered message is spam or not using Multinominal Naive Bayes classifier

In [1]:
#import necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

Read the dataset

In [2]:
df=pd.read_csv(r'file:///C:\Users\dell\Downloads\pandas\emailspam.csv')

In [3]:
df2=pd.read_csv(r'file:///C:\Users\dell\Downloads\pandas\emailspam.csv')

In [4]:
df=pd.concat([df,df2])

Finding the how many rows & columns present in dataset

Let's take overview of our data

In [5]:
df.shape

(11144, 2)

So there are 11144 rows & 2 columns present in dataset

In [6]:
df.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

Finding there is any missing value is present

In [7]:
df.isnull().sum()

Category    0
Message     0
dtype: int64

So there is no missing values present in both clumns

In [8]:
df['Category'].value_counts()

ham     9650
spam    1494
Name: Category, dtype: int64

'ham' represent this is no spam message  & spam represent this is not a spam message

So we get how many spam messages & how many Non spam messages are present
Spam messages are 747 & non spam messages are 4825

In [9]:
df['Category'].unique()

array(['ham', 'spam'], dtype=object)

In [10]:
df['Message'].nunique()

5157

Create a independant variable as 'x'

In [11]:
X=df['Message']

Performing LabelEncoding on Category Column

Which is our target attribute Which has denoted by 'y' variable

In [12]:
from sklearn.preprocessing import LabelEncoder
lb=LabelEncoder()
y=lb.fit_transform(df['Category'])

We successfully conver categorical values into numerical

In [13]:
y

array([0, 0, 1, ..., 0, 0, 0])

Spliting the dataset into training set & testing set using 80-20 split method

In [14]:
from sklearn.model_selection import train_test_split
Xtrain, Xtest, ytrain, ytest = train_test_split(X,y,test_size=0.20, random_state=0)

Finding the shape of training set & testing set

In [15]:
Xtest.shape,Xtrain.shape,ytest.shape,ytrain.shape

((2229,), (8915,), (2229,), (8915,))

To convert text into vectors let's use 'CountVectorizer' algorithm

So let's create pipeline to feed the data to model

Now let's see the accuracies of different different classification algorithms

In [16]:

from sklearn.feature_extraction.text import CountVectorizer
#from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
classifier=Pipeline([
    ('vectorizer',CountVectorizer()),
    ('classifier',LogisticRegression())
])
#from xgboost import XGBClassifier
#classifier = LogisticRegression() 
classifier.fit(Xtrain, ytrain)

# Predicting the Test set results
y_pred = classifier.predict(Xtest)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(ytest, y_pred)
print(cm)#error rate 0.15

from sklearn.metrics import classification_report
cr = classification_report(ytest, y_pred)
print(cr)

from sklearn.metrics import accuracy_score
ac = accuracy_score(ytest, y_pred)
print(ac)

[[1917    0]
 [  17  295]]
              precision    recall  f1-score   support

           0       0.99      1.00      1.00      1917
           1       1.00      0.95      0.97       312

    accuracy                           0.99      2229
   macro avg       1.00      0.97      0.98      2229
weighted avg       0.99      0.99      0.99      2229

0.9923732615522656


In [17]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn import svm
from sklearn.pipeline import Pipeline
classifier=Pipeline([
    ('vectorizer',CountVectorizer()),
    ('classifier',svm.SVC())
])
#from xgboost import XGBClassifier
#classifier = LogisticRegression() 
classifier.fit(Xtrain, ytrain)

# Predicting the Test set results
y_pred = classifier.predict(Xtest)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(ytest, y_pred)
print(cm)#error rate 0.15

from sklearn.metrics import classification_report
cr = classification_report(ytest, y_pred)
print(cr)

from sklearn.metrics import accuracy_score
ac = accuracy_score(ytest, y_pred)
print(ac)

[[1917    0]
 [  18  294]]
              precision    recall  f1-score   support

           0       0.99      1.00      1.00      1917
           1       1.00      0.94      0.97       312

    accuracy                           0.99      2229
   macro avg       1.00      0.97      0.98      2229
weighted avg       0.99      0.99      0.99      2229

0.9919246298788694


In [18]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline
classifier=Pipeline([
    ('vectorizer',CountVectorizer()),
    ('classifier',KNeighborsClassifier())
])
#from xgboost import XGBClassifier
#classifier = LogisticRegression() 
classifier.fit(Xtrain, ytrain)

# Predicting the Test set results
y_pred = classifier.predict(Xtest)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(ytest, y_pred)
print(cm)#error rate 0.15

from sklearn.metrics import classification_report
cr = classification_report(ytest, y_pred)
print(cr)

from sklearn.metrics import accuracy_score
ac = accuracy_score(ytest, y_pred)
print(ac)

[[1916    1]
 [ 124  188]]
              precision    recall  f1-score   support

           0       0.94      1.00      0.97      1917
           1       0.99      0.60      0.75       312

    accuracy                           0.94      2229
   macro avg       0.97      0.80      0.86      2229
weighted avg       0.95      0.94      0.94      2229

0.9439210408254822


In [19]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.tree import DecisionTreeClassifier
from sklearn.pipeline import Pipeline
classifier=Pipeline([
    ('vectorizer',CountVectorizer()),
    ('classifier',DecisionTreeClassifier())
])
#from xgboost import XGBClassifier
#classifier = LogisticRegression() 
classifier.fit(Xtrain, ytrain)

# Predicting the Test set results
y_pred = classifier.predict(Xtest)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(ytest, y_pred)
print(cm)#error rate 0.15

from sklearn.metrics import classification_report
cr = classification_report(ytest, y_pred)
print(cr)

from sklearn.metrics import accuracy_score
ac = accuracy_score(ytest, y_pred)
print(ac)

[[1913    4]
 [  12  300]]
              precision    recall  f1-score   support

           0       0.99      1.00      1.00      1917
           1       0.99      0.96      0.97       312

    accuracy                           0.99      2229
   macro avg       0.99      0.98      0.98      2229
weighted avg       0.99      0.99      0.99      2229

0.9928218932256617


In [20]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline
classifier=Pipeline([
    ('vectorizer',CountVectorizer()),
    ('classifier',RandomForestClassifier())
])
#from xgboost import XGBClassifier
#classifier = LogisticRegression() 
classifier.fit(Xtrain, ytrain)

# Predicting the Test set results
y_pred = classifier.predict(Xtest)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(ytest, y_pred)
print(cm)#error rate 0.15

from sklearn.metrics import classification_report
cr = classification_report(ytest, y_pred)
print(cr)

from sklearn.metrics import accuracy_score
ac = accuracy_score(ytest, y_pred)
print(ac)

[[1917    0]
 [  14  298]]
              precision    recall  f1-score   support

           0       0.99      1.00      1.00      1917
           1       1.00      0.96      0.98       312

    accuracy                           0.99      2229
   macro avg       1.00      0.98      0.99      2229
weighted avg       0.99      0.99      0.99      2229

0.993719156572454


In [21]:
#CounterVectorizer Convert the text into matrics
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline
clssifier=Pipeline([
    ('vectorizer',CountVectorizer()),
    ('nb',MultinomialNB())
])

To train the dataset I have used Multinominal Naive Bayes Model

Now there is no problem i can fit the model

In [22]:
clssifier.fit(Xtrain,ytrain)

Pipeline(steps=[('vectorizer', CountVectorizer()), ('nb', MultinomialNB())])

apply predict method on the Xtest dataset

In [23]:
ypred=clssifier.predict(Xtest)

In [24]:
ypred

array([0, 0, 0, ..., 1, 0, 0])

Bias

In [25]:
bias=clssifier.score(Xtrain,ytrain)
bias

0.993157599551318

Variance

In [26]:
variance=clssifier.score(Xtest,ytest)
variance

0.9896814715118888

Creating a confusion matrix to find error rate or accuracy 

In [27]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(ytest, ypred)
print(cm)

[[1909    8]
 [  15  297]]


So the Accuracy is 0.99 & Error rate is 0.1

#create a classification report

In [28]:
from sklearn.metrics import classification_report
cr = classification_report(ytest, ypred)
print(cr)

              precision    recall  f1-score   support

           0       0.99      1.00      0.99      1917
           1       0.97      0.95      0.96       312

    accuracy                           0.99      2229
   macro avg       0.98      0.97      0.98      2229
weighted avg       0.99      0.99      0.99      2229



In [29]:
from sklearn.metrics import accuracy_score
ac = accuracy_score(ytest, ypred)
print(ac)

0.9896814715118888


Lets take demo of our model

Prediction 0 means this not a spam message & 1 means this is spam message

Let's i enter some Spam messages from dataset

In [30]:
emails=[
    'WINNER!! As a valued network customer you have been selected to receivea £900 prize reward! To claim call 09061701461. Claim code KL341. Valid 12 hours only.'
]

In [31]:
res=clssifier.predict(emails)

In [32]:
res

array([1])

If i look this message in dataset i found that this is Spam message & it has given me correct output.

In [33]:
mg=['Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&Cs apply 08452810075over18']

In [34]:
clssifier.predict(mg)

array([1])

In [35]:
mesg=['Hello how are you?']

In [36]:
clssifier.predict(mesg)

array([0])

So the our model is ready to predict spam messages or emails

Export the model

In [37]:
import joblib
model=joblib.dump(classifier,'spam_clff.pkl')
model = joblib.load("spam_clff.pkl")